pkg/services/ngalert/state/historian/core.go

```diff
@@ -34,6 +34,24 @@ func shouldRecord(transition state.StateTransition) bool {
 	return true
 }
 
+func shouldRecordAnnotation(rule history_model.RuleMeta, t state.StateTransition) bool {
+	if !shouldRecord(t) {
+		return false
+	}
+
+	// If user mapped NoData->OK, do not record transitions between Normal and Normal(NoData) and back.
+	if rule.NoDataState == models.OK &&
+		t.State.State == eval.Normal &&
+		t.PreviousState == eval.Normal {
+		if (t.State.StateReason == "" && t.PreviousStateReason == models.StateReasonNoData) ||
+			(t.State.StateReason == models.StateReasonNoData && t.PreviousStateReason == "") {
+			return false
+		}
+	}
```

Comment:
I think it can be simplified and check for `rule.NoDataState == models.Ok` being dropped. I think this leaks the knowledge about mapping outside state management domain.

Reply:
Currently yes, the only vector for `Normal (NoData)` is mapping NoData to OK.

To your point, the check for `Normal (Paused) -> Normal` doesn't use `rule.IsPaused` because `StateReasonPaused` identifies the rule state as paused. Likewise, `StateReasonNoData` does the same for a NoData eval. If this model breaks down a bit it would be because the semantics of `StateReasonNoData` have changed, and that change should include making sure code using the constant is aware of the new semantics.